<!DOCTYPE html><html><head><meta charset="utf-8"><style>@font-face {
  font-family: octicons-anchor;
  src: url(data:font/woff;charset=utf-8;base64,d09GRgABAAAAAAYcAA0AAAAACjQAAQAAAAAAAAAAAAAAAAAAAAAAAAAAAABGRlRNAAABMAAAABwAAAAca8vGTk9TLzIAAAFMAAAARAAAAFZG1VHVY21hcAAAAZAAAAA+AAABQgAP9AdjdnQgAAAB0AAAAAQAAAAEACICiGdhc3AAAAHUAAAACAAAAAj//wADZ2x5ZgAAAdwAAADRAAABEKyikaNoZWFkAAACsAAAAC0AAAA2AtXoA2hoZWEAAALgAAAAHAAAACQHngNFaG10eAAAAvwAAAAQAAAAEAwAACJsb2NhAAADDAAAAAoAAAAKALIAVG1heHAAAAMYAAAAHwAAACABEAB2bmFtZQAAAzgAAALBAAAFu3I9x/Nwb3N0AAAF/AAAAB0AAAAvaoFvbwAAAAEAAAAAzBdyYwAAAADP2IQvAAAAAM/bz7t4nGNgZGFgnMDAysDB1Ml0hoGBoR9CM75mMGLkYGBgYmBlZsAKAtJcUxgcPsR8iGF2+O/AEMPsznAYKMwIkgMA5REMOXicY2BgYGaAYBkGRgYQsAHyGMF8FgYFIM0ChED+h5j//yEk/3KoSgZGNgYYk4GRCUgwMaACRoZhDwCs7QgGAAAAIgKIAAAAAf//AAJ4nHWMMQrCQBBF/0zWrCCIKUQsTDCL2EXMohYGSSmorScInsRGL2DOYJe0Ntp7BK+gJ1BxF1stZvjz/v8DRghQzEc4kIgKwiAppcA9LtzKLSkdNhKFY3HF4lK69ExKslx7Xa+vPRVS43G98vG1DnkDMIBUgFN0MDXflU8tbaZOUkXUH0+U27RoRpOIyCKjbMCVejwypzJJG4jIwb43rfl6wbwanocrJm9XFYfskuVC5K/TPyczNU7b84CXcbxks1Un6H6tLH9vf2LRnn8Ax7A5WQAAAHicY2BkYGAA4teL1+yI57f5ysDNwgAC529f0kOmWRiYVgEpDgYmEA8AUzEKsQAAAHicY2BkYGB2+O/AEMPCAAJAkpEBFbAAADgKAe0EAAAiAAAAAAQAAAAEAAAAAAAAKgAqACoAiAAAeJxjYGRgYGBhsGFgYgABEMkFhAwM/xn0QAIAD6YBhwB4nI1Ty07cMBS9QwKlQapQW3VXySvEqDCZGbGaHULiIQ1FKgjWMxknMfLEke2A+IJu+wntrt/QbVf9gG75jK577Lg8K1qQPCfnnnt8fX1NRC/pmjrk/zprC+8D7tBy9DHgBXoWfQ44Av8t4Bj4Z8CLtBL9CniJluPXASf0Lm4CXqFX8Q84dOLnMB17N4c7tBo1AS/Qi+hTwBH4rwHHwN8DXqQ30XXAS7QaLwSc0Gn8NuAVWou/gFmnjLrEaEh9GmDdDGgL3B4JsrRPDU2hTOiMSuJUIdKQQayiAth69r6akSSFqIJuA19TrzCIaY8sIoxyrNIrL//pw7A2iMygkX5vDj+G+kuoLdX4GlGK/8Lnlz6/h9MpmoO9rafrz7ILXEHHaAx95s9lsI7AHNMBWEZHULnfAXwG9/ZqdzLI08iuwRloXE8kfhXYAvE23+23DU3t626rbs8/8adv+9DWknsHp3E17oCf+Z48rvEQNZ78paYM38qfk3v/u3l3u3GXN2Dmvmvpf1Srwk3pB/VSsp512bA/GG5i2WJ7wu430yQ5K3nFGiOqgtmSB5pJVSizwaacmUZzZhXLlZTq8qGGFY2YcSkqbth6aW1tRmlaCFs2016m5qn36SbJrqosG4uMV4aP2PHBmB3tjtmgN2izkGQyLWprekbIntJFing32a5rKWCN/SdSoga45EJykyQ7asZvHQ8PTm6cslIpwyeyjbVltNikc2HTR7YKh9LBl9DADC0U/jLcBZDKrMhUBfQBvXRzLtFtjU9eNHKin0x5InTqb8lNpfKv1s1xHzTXRqgKzek/mb7nB8RZTCDhGEX3kK/8Q75AmUM/eLkfA+0Hi908Kx4eNsMgudg5GLdRD7a84npi+YxNr5i5KIbW5izXas7cHXIMAau1OueZhfj+cOcP3P8MNIWLyYOBuxL6DRylJ4cAAAB4nGNgYoAALjDJyIAOWMCiTIxMLDmZedkABtIBygAAAA==) format('woff');
}

* {
    box-sizing: border-box;
}

body {
    width: 980px;
    margin-right: auto;
    margin-left: auto;
}

body .markdown-body {
    padding: 45px;
    border: 1px solid #ddd;
    border-radius: 3px;
    word-wrap: break-word;
}

pre {
    font: 12px Consolas, "Liberation Mono", Menlo, Courier, monospace;
}

.markdown-body {
  -webkit-text-size-adjust: 100%;
  text-size-adjust: 100%;
  color: #333;
  font-family: "Helvetica Neue", Helvetica, "Segoe UI", Arial, freesans, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
  font-size: 16px;
  line-height: 1.6;
  word-wrap: break-word;
}

.markdown-body a {
  background-color: transparent;
}

.markdown-body a:active,
.markdown-body a:hover {
  outline: 0;
}

.markdown-body strong {
  font-weight: bold;
}

.markdown-body h1 {
  font-size: 2em;
  margin: 0.67em 0;
}

.markdown-body img {
  border: 0;
}

.markdown-body hr {
  box-sizing: content-box;
  height: 0;
}

.markdown-body pre {
  overflow: auto;
}

.markdown-body code,
.markdown-body kbd,
.markdown-body pre {
  font-family: monospace, monospace;
  font-size: 1em;
}

.markdown-body input {
  color: inherit;
  font: inherit;
  margin: 0;
}

.markdown-body html input[disabled] {
  cursor: default;
}

.markdown-body input {
  line-height: normal;
}

.markdown-body input[type="checkbox"] {
  box-sizing: border-box;
  padding: 0;
}

.markdown-body table {
  border-collapse: collapse;
  border-spacing: 0;
}

.markdown-body td,
.markdown-body th {
  padding: 0;
}

.markdown-body input {
  font: 13px / 1.4 Helvetica, arial, nimbussansl, liberationsans, freesans, clean, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
}

.markdown-body a {
  color: #4078c0;
  text-decoration: none;
}

.markdown-body a:hover,
.markdown-body a:active {
  text-decoration: underline;
}

.markdown-body hr {
  height: 0;
  margin: 15px 0;
  overflow: hidden;
  background: transparent;
  border: 0;
  border-bottom: 1px solid #ddd;
}

.markdown-body hr:before {
  display: table;
  content: "";
}

.markdown-body hr:after {
  display: table;
  clear: both;
  content: "";
}

.markdown-body h1,
.markdown-body h2,
.markdown-body h3,
.markdown-body h4,
.markdown-body h5,
.markdown-body h6 {
  margin-top: 15px;
  margin-bottom: 15px;
  line-height: 1.1;
}

.markdown-body h1 {
  font-size: 30px;
}

.markdown-body h2 {
  font-size: 21px;
}

.markdown-body h3 {
  font-size: 16px;
}

.markdown-body h4 {
  font-size: 14px;
}

.markdown-body h5 {
  font-size: 12px;
}

.markdown-body h6 {
  font-size: 11px;
}

.markdown-body blockquote {
  margin: 0;
}

.markdown-body ul,
.markdown-body ol {
  padding: 0;
  margin-top: 0;
  margin-bottom: 0;
}

.markdown-body ol ol,
.markdown-body ul ol {
  list-style-type: lower-roman;
}

.markdown-body ul ul ol,
.markdown-body ul ol ol,
.markdown-body ol ul ol,
.markdown-body ol ol ol {
  list-style-type: lower-alpha;
}

.markdown-body dd {
  margin-left: 0;
}

.markdown-body code {
  font-family: Consolas, "Liberation Mono", Menlo, Courier, monospace;
  font-size: 12px;
}

.markdown-body pre {
  margin-top: 0;
  margin-bottom: 0;
  font: 12px Consolas, "Liberation Mono", Menlo, Courier, monospace;
}

.markdown-body .select::-ms-expand {
  opacity: 0;
}

.markdown-body .octicon {
  font: normal normal normal 16px/1 octicons-anchor;
  display: inline-block;
  text-decoration: none;
  text-rendering: auto;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.markdown-body .octicon-link:before {
  content: '\f05c';
}

.markdown-body:before {
  display: table;
  content: "";
}

.markdown-body:after {
  display: table;
  clear: both;
  content: "";
}

.markdown-body>*:first-child {
  margin-top: 0 !important;
}

.markdown-body>*:last-child {
  margin-bottom: 0 !important;
}

.markdown-body a:not([href]) {
  color: inherit;
  text-decoration: none;
}

.markdown-body .anchor {
  display: inline-block;
  padding-right: 2px;
  margin-left: -18px;
}

.markdown-body .anchor:focus {
  outline: none;
}

.markdown-body h1,
.markdown-body h2,
.markdown-body h3,
.markdown-body h4,
.markdown-body h5,
.markdown-body h6 {
  margin-top: 1em;
  margin-bottom: 16px;
  font-weight: bold;
  line-height: 1.4;
}

.markdown-body h1 .octicon-link,
.markdown-body h2 .octicon-link,
.markdown-body h3 .octicon-link,
.markdown-body h4 .octicon-link,
.markdown-body h5 .octicon-link,
.markdown-body h6 .octicon-link {
  color: #000;
  vertical-align: middle;
  visibility: hidden;
}

.markdown-body h1:hover .anchor,
.markdown-body h2:hover .anchor,
.markdown-body h3:hover .anchor,
.markdown-body h4:hover .anchor,
.markdown-body h5:hover .anchor,
.markdown-body h6:hover .anchor {
  text-decoration: none;
}

.markdown-body h1:hover .anchor .octicon-link,
.markdown-body h2:hover .anchor .octicon-link,
.markdown-body h3:hover .anchor .octicon-link,
.markdown-body h4:hover .anchor .octicon-link,
.markdown-body h5:hover .anchor .octicon-link,
.markdown-body h6:hover .anchor .octicon-link {
  visibility: visible;
}

.markdown-body h1 {
  padding-bottom: 0.3em;
  font-size: 2.25em;
  line-height: 1.2;
  border-bottom: 1px solid #eee;
}

.markdown-body h1 .anchor {
  line-height: 1;
}

.markdown-body h2 {
  padding-bottom: 0.3em;
  font-size: 1.75em;
  line-height: 1.225;
  border-bottom: 1px solid #eee;
}

.markdown-body h2 .anchor {
  line-height: 1;
}

.markdown-body h3 {
  font-size: 1.5em;
  line-height: 1.43;
}

.markdown-body h3 .anchor {
  line-height: 1.2;
}

.markdown-body h4 {
  font-size: 1.25em;
}

.markdown-body h4 .anchor {
  line-height: 1.2;
}

.markdown-body h5 {
  font-size: 1em;
}

.markdown-body h5 .anchor {
  line-height: 1.1;
}

.markdown-body h6 {
  font-size: 1em;
  color: #777;
}

.markdown-body h6 .anchor {
  line-height: 1.1;
}

.markdown-body p,
.markdown-body blockquote,
.markdown-body ul,
.markdown-body ol,
.markdown-body dl,
.markdown-body table,
.markdown-body pre {
  margin-top: 0;
  margin-bottom: 16px;
}

.markdown-body hr {
  height: 4px;
  padding: 0;
  margin: 16px 0;
  background-color: #e7e7e7;
  border: 0 none;
}

.markdown-body ul,
.markdown-body ol {
  padding-left: 2em;
}

.markdown-body ul ul,
.markdown-body ul ol,
.markdown-body ol ol,
.markdown-body ol ul {
  margin-top: 0;
  margin-bottom: 0;
}

.markdown-body li>p {
  margin-top: 16px;
}

.markdown-body dl {
  padding: 0;
}

.markdown-body dl dt {
  padding: 0;
  margin-top: 16px;
  font-size: 1em;
  font-style: italic;
  font-weight: bold;
}

.markdown-body dl dd {
  padding: 0 16px;
  margin-bottom: 16px;
}

.markdown-body blockquote {
  padding: 0 15px;
  color: #777;
  border-left: 4px solid #ddd;
}

.markdown-body blockquote>:first-child {
  margin-top: 0;
}

.markdown-body blockquote>:last-child {
  margin-bottom: 0;
}

.markdown-body table {
  display: block;
  width: 100%;
  overflow: auto;
  word-break: normal;
  word-break: keep-all;
}

.markdown-body table th {
  font-weight: bold;
}

.markdown-body table th,
.markdown-body table td {
  padding: 6px 13px;
  border: 1px solid #ddd;
}

.markdown-body table tr {
  background-color: #fff;
  border-top: 1px solid #ccc;
}

.markdown-body table tr:nth-child(2n) {
  background-color: #f8f8f8;
}

.markdown-body img {
  max-width: 100%;
  box-sizing: content-box;
  background-color: #fff;
}

.markdown-body code {
  padding: 0;
  padding-top: 0.2em;
  padding-bottom: 0.2em;
  margin: 0;
  font-size: 85%;
  background-color: rgba(0,0,0,0.04);
  border-radius: 3px;
}

.markdown-body code:before,
.markdown-body code:after {
  letter-spacing: -0.2em;
  content: "\00a0";
}

.markdown-body pre>code {
  padding: 0;
  margin: 0;
  font-size: 100%;
  word-break: normal;
  white-space: pre;
  background: transparent;
  border: 0;
}

.markdown-body .highlight {
  margin-bottom: 16px;
}

.markdown-body .highlight pre,
.markdown-body pre {
  padding: 16px;
  overflow: auto;
  font-size: 85%;
  line-height: 1.45;
  background-color: #f7f7f7;
  border-radius: 3px;
}

.markdown-body .highlight pre {
  margin-bottom: 0;
  word-break: normal;
}

.markdown-body pre {
  word-wrap: normal;
}

.markdown-body pre code {
  display: inline;
  max-width: initial;
  padding: 0;
  margin: 0;
  overflow: initial;
  line-height: inherit;
  word-wrap: normal;
  background-color: transparent;
  border: 0;
}

.markdown-body pre code:before,
.markdown-body pre code:after {
  content: normal;
}

.markdown-body kbd {
  display: inline-block;
  padding: 3px 5px;
  font-size: 11px;
  line-height: 10px;
  color: #555;
  vertical-align: middle;
  background-color: #fcfcfc;
  border: solid 1px #ccc;
  border-bottom-color: #bbb;
  border-radius: 3px;
  box-shadow: inset 0 -1px 0 #bbb;
}

.markdown-body .pl-c {
  color: #969896;
}

.markdown-body .pl-c1,
.markdown-body .pl-s .pl-v {
  color: #0086b3;
}

.markdown-body .pl-e,
.markdown-body .pl-en {
  color: #795da3;
}

.markdown-body .pl-s .pl-s1,
.markdown-body .pl-smi {
  color: #333;
}

.markdown-body .pl-ent {
  color: #63a35c;
}

.markdown-body .pl-k {
  color: #a71d5d;
}

.markdown-body .pl-pds,
.markdown-body .pl-s,
.markdown-body .pl-s .pl-pse .pl-s1,
.markdown-body .pl-sr,
.markdown-body .pl-sr .pl-cce,
.markdown-body .pl-sr .pl-sra,
.markdown-body .pl-sr .pl-sre {
  color: #183691;
}

.markdown-body .pl-v {
  color: #ed6a43;
}

.markdown-body .pl-id {
  color: #b52a1d;
}

.markdown-body .pl-ii {
  background-color: #b52a1d;
  color: #f8f8f8;
}

.markdown-body .pl-sr .pl-cce {
  color: #63a35c;
  font-weight: bold;
}

.markdown-body .pl-ml {
  color: #693a17;
}

.markdown-body .pl-mh,
.markdown-body .pl-mh .pl-en,
.markdown-body .pl-ms {
  color: #1d3e81;
  font-weight: bold;
}

.markdown-body .pl-mq {
  color: #008080;
}

.markdown-body .pl-mi {
  color: #333;
  font-style: italic;
}

.markdown-body .pl-mb {
  color: #333;
  font-weight: bold;
}

.markdown-body .pl-md {
  background-color: #ffecec;
  color: #bd2c00;
}

.markdown-body .pl-mi1 {
  background-color: #eaffea;
  color: #55a532;
}

.markdown-body .pl-mdr {
  color: #795da3;
  font-weight: bold;
}

.markdown-body .pl-mo {
  color: #1d3e81;
}

.markdown-body kbd {
  display: inline-block;
  padding: 3px 5px;
  font: 11px Consolas, "Liberation Mono", Menlo, Courier, monospace;
  line-height: 10px;
  color: #555;
  vertical-align: middle;
  background-color: #fcfcfc;
  border: solid 1px #ccc;
  border-bottom-color: #bbb;
  border-radius: 3px;
  box-shadow: inset 0 -1px 0 #bbb;
}

.markdown-body .plan-price-unit {
  color: #767676;
  font-weight: normal;
}

.markdown-body .task-list-item {
  list-style-type: none;
}

.markdown-body .task-list-item+.task-list-item {
  margin-top: 3px;
}

.markdown-body .task-list-item input {
  margin: 0 0.35em 0.25em -1.6em;
  vertical-align: middle;
}

.markdown-body .plan-choice {
  padding: 15px;
  padding-left: 40px;
  display: block;
  border: 1px solid #e0e0e0;
  position: relative;
  font-weight: normal;
  background-color: #fafafa;
}

.markdown-body .plan-choice.open {
  background-color: #fff;
}

.markdown-body .plan-choice.open .plan-choice-seat-breakdown {
  display: block;
}

.markdown-body .plan-choice-free {
  border-radius: 3px 3px 0 0;
}

.markdown-body .plan-choice-paid {
  border-radius: 0 0 3px 3px;
  border-top: 0;
  margin-bottom: 20px;
}

.markdown-body .plan-choice-radio {
  position: absolute;
  left: 15px;
  top: 18px;
}

.markdown-body .plan-choice-exp {
  color: #999;
  font-size: 12px;
  margin-top: 5px;
}

.markdown-body .plan-choice-seat-breakdown {
  margin-top: 10px;
  display: none;
}

.markdown-body :checked+.radio-label {
  z-index: 1;
  position: relative;
  border-color: #4078c0;
}
</style><title>README</title></head><body><article class="markdown-body"><h1>
<a id="user-content-bcsuite" class="anchor" href="#bcsuite" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>BCSuite</h1>
<p>BCSuite contains a series of scripts, that constitute a DNA BarCode filter/design pipeline.</p>
<p>This document explains in detail the current pipeline for barcode filter/design. The barcodes are selected based on three features: reference genome homology, self-dimerization free energy, and hetero-dimerization free energy.</p>
<h2>
<a id="user-content-pipeline" class="anchor" href="#pipeline" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Pipeline</h2>
<ol>
<li>Sequence generation.</li>
<li>
<a href="https://genome.ucsc.edu/cgi-bin/hgBlat?command=start">BLAT</a> against reference genome.</li>
<li>Merge BLAT output into single file.</li>
<li>Split merged BLAT output into chunks.</li>
<li>Convert chunks into blat recap output.</li>
<li>Calculate self-dimerization free energy (sdfe) for every barcode candidate.</li>
<li>Filter based on maximum homology and sdfe.</li>
<li>Calculate pair-wise hetero-dimerization free energy (hdfe) for the barcode candidates selected from the filter (7).</li>
<li>Filter based on hdfe.</li>
<li>Select barcode set with <a href="https://www.cs.purdue.edu/homes/dgleich/codes/maxcliques/README.html">PMC algorithm</a>.</li>
</ol>
<h3>
<a id="user-content-1-sequence-generation" class="anchor" href="#1-sequence-generation" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>1. Sequence generation</h3>
<p>Of course, we need to produce a set of barcode candidate to be filtered and properly selected. This set can be generated in a variety of ways. One example, would be selecting all the 20 characters long substring of a set of 25-mers.</p>
<p>Specifically, Xu et al. published a list of 240,000 25-mers, that can be used to generate a list of 1440000 20-mers.</p>
<h3>
<a id="user-content-2-blat" class="anchor" href="#2-blat" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>2. <a href="https://genome.ucsc.edu/cgi-bin/hgBlat?command=start">BLAT</a>
</h3>
<p>Depending on the application, it might be important for the barcodes to be genome orthogonal, i.e., being absent from the genome. In other words, the homology between a barcode and the genome should be lower than a given threshold (e.g., 60%). This can be achieved using BLAT to align each barcode to the reference genome, and then analyze the output.</p>
<p>Specifically, if the barcode is short (i.e., ~20 nt), the following parameters should be used:</p>
<pre><code>echo -e "&gt; mer_XXXX\\nACTAG....ACTG" | blat -tileSize=6 -stepSize=1
-minMatch=1 -oneOff=1 -minScore=0 -minIdentity=0 -maxGap=0 -repMatch=131071
-noHead list.fa stdin &gt; mer_xxxx.psl
</code></pre>
<ul>
<li>
<strong>input</strong>: BLAT input should be a fasta-like string, containing the header and the sequence for a single barcode.</li>
<li>
<strong>tileSize</strong>: size of match that triggers an alignment. Default is 11 for DNA, usually set between 8 an 12.</li>
<li>
<strong>stepSize</strong>: spacing between tiles. Default is tileSize.</li>
<li>
<strong>minMatch</strong>: number of tile matches. usually set from 2 to 4. Default is 2 for nucleotide.</li>
<li>
<strong>oneOff</strong>: if set to 1, this allows one mismatch in tile and still triggers an alignment. Default is 0.</li>
<li>
<strong>minScore</strong>: sets minimum score. The matches minus the mismatches minus a gap penalty. Default is 30.</li>
<li>
<strong>minIdentity</strong>: sets minimum identity in percent. Default is 90 for nucleotide seraches.</li>
<li>
<strong>maxGap</strong>: size of maximum gap between tiles in a clump. Usually set from 0 to 3. Default is 2. Only relevant for minMatch &gt; 1.</li>
<li>
<strong>repMatch</strong>: number of repetitions of a tile allowed before it is marked as overused. Typically is 256 for tileSize 12, 1024 for tileSize 11, 4096 for tileSize 10... Default is 1024.Also affected by stepSize: when stepSize is halved, repMatch is doubled to compensate.</li>
<li>
<strong>noHead</strong>: remove header from the output.</li>
<li>
<strong>list.fa</strong>: reference genome.</li>
<li>
<strong>stdin</strong>: query (from stdin).</li>
<li>
<strong>output</strong>: the output will be a PSL file [<a href="http://www.ensembl.org/info/website/upload/psl.html">PSL format definition</a>].</li>
</ul>
<h3>
<a id="user-content-3-merge" class="anchor" href="#3-merge" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>3. Merge</h3>
<p>To convert the output of BLAT in a more readable format (here addressed as <em>recap</em>), we first need to merge all the single barcode PSL files into a single file.</p>
<p>Usually, this can be easily achieved with a simple command line:</p>
<pre><code>cat psl/* &gt; merged_psl.psl
</code></pre>
<p>But, when the number of single barcode PSL file is too large, it is necessary to use the <code>psl_merge.sh</code>. By going in the folder containing the PSL files, just run:</p>
<pre><code>ls | xargs -n 1000 -P 1 psl_merge.sh
</code></pre>
<p>This will generate the <code>../merged_psl.psl</code> file.</p>
<h3>
<a id="user-content-4-split" class="anchor" href="#4-split" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>4. Split</h3>
<p>Depending on the amount of available RAM on your machine, it might be difficult to operate on the <code>merged_psl.psl</code> file when many barcodes are being analyzed at the same time. We suggest then to split the file into smaller chunks using the following command line:</p>
<pre><code>split -l 2GB  merged_psl.psl split/psl.
</code></pre>
<p>This command will split the <code>merged_psl.psl</code> file into 2GB-sized chunks of name <code>psl.aa</code>, <code>psl.ab</code>,... in the <code>split/</code> folder.</p>
<h3>
<a id="user-content-5-recap" class="anchor" href="#5-recap" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>5. Recap</h3>
<p>The <em>recap</em> format is useful to understand the output of the BLAT tool. Specifically, it contains one barcode (sequence) per row with the following columns:</p>
<ul>
<li>
<strong>i</strong>: barcode ID number.</li>
<li>
<strong>qName</strong>: barcode ID name.</li>
<li>
<strong>k</strong>: barcode length in nt.</li>
<li>
<strong>seq</strong>: barcode sequence.</li>
<li>
<strong>hX</strong>: number of genomic loci with X homology with the barcode.</li>
<li>
<strong>maxHomology</strong>: highest homology with a genomic locus (highest X with <code>hX != 0</code>).</li>
</ul>
<p>The script that converts the merged PSL file into the <em>recap</em> format is called <code>recap_blat.R</code>.</p>
<pre><code>usage: recap_blat.R [--help] seq_dir psl_dir out_dir
       [--seq SEQ] [--psl PSL] [--out OUT] [--nst NST]
       [--ncores NCORES] [--suff SUFF]

Recaps BLAT output.

positional arguments:
  seq_dir           Directory containing the sequence file.
  psl_dir           Directory containing the psl file.
  out_dir           Output directory.

flags:
  -h, --help        show this help message and exit

optional arguments:
  -s, --seq SEQ     Sequence file name. [default: bc.noHeader.fa]
  -o, --out OUT     Output file name. [default: merged_table.tsv]
  -p, --psl  PSL    PSL file name. [default: merged_psl.psl]
  -n, --nst NST     Number of sub tables. [default: 1000]
  --ncores NCORES   Number of cores for parallelization. [default: 30]
  --suff SUFF       Output suffix. [default: ]
</code></pre>
<h5>
<a id="user-content-recap-after-splitting" class="anchor" href="#recap-after-splitting" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Recap after splitting</h5>
<p>If the split step (4) was performed, the <code>recap_blat.R</code> script function should be run on every split PSL file. This can be easily achieved by running the <code>recap_split.sh</code> script.</p>
<pre><code>usage: recap_split.sh -i psl_dir -o out_dir -s seq_dir
       [-f seq_file][-t threads][-p path]
 
Run the recap_blat.R script for every split PSL file
in the specified psl (-i) directory.

Mandatory arguments:
 -i psl_dir    Directory containing the split PSL files.
 -s seq_dir    Directory containing the seq_file (-f).
 -o out_dir    Output directory.

Optional arguments:
 -f seq_file   Name of the sequence file in seq_dir (-s).
 -t threads    Number of threads for parallelization.
 -p path Path  to recap_blat.R script. Default './recap_blat.R'
</code></pre>
<p>Then, merge the converted split recap tables with the <code>recap_merge.R</code> script.</p>
<pre><code>usage: recap_merge.R [--help] recap_dir out_dir [--out OUT]

Merge BLAT recaps.

positional arguments:
  recap_dir         Directory containing the recap files.
  out_dir           Output directory.

flags:
  -h, --help        show this help message and exit

optional arguments:
  -o, --out OUT     Output file name. [default: merged_table.tsv]
</code></pre>
<h3>
<a id="user-content-6-calculate-sdfe" class="anchor" href="#6-calculate-sdfe" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>6. Calculate sdfe</h3>
<p>If needed, it is also possible to add a column with the calculated self-dimerization free energy (sdfe) to the recap table, using the <code>sdfe_calc.R</code> script.</p>
<pre><code>usage: sdfe_calc.R [--help] rec_dir out_dir
       [--rec REC] [--out OUT] [--ncores NCORES] [--suff SUFF]

Calculate self dimerization free energy.

positional arguments:
  rec_dir               Directory containing the recap table.
  out_dir               Output directory.

flags:
  -h, --help            show this help message and exit

optional arguments:
  -r, --rec REC         Recap table file name. [default: merged_table.tsv]
  -o, --out OUT         Output file name. [default: merged_table.sdfe.tsv]
  -n, --ncores NCORES   Number of cores for parallelization. [default: 30]
  -s, --suff SUFF       Output suffix. [default: ]
</code></pre>
<h3>
<a id="user-content-7-filter-maxhomology--sdfe" class="anchor" href="#7-filter-maxhomology--sdfe" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>7. Filter (maxHomology &amp; sdfe)</h3>
<p>Then, the recap table with sdfe column should be filtered based on sdfe and maxHomology with the <code>sdfe_filter.R</code> script. Use the output of the previous step (6) as the input (fin). The default SDFE threshold is set at -5 kcal/mol while the maxHomology at 75% of the sequence.</p>
<pre><code>usage: sdfe_filter.R [--help] in_dir fin out_dir
       [--out OUT] [--homo_thr HOMO_THR] [--sdfe_thr SDFE_THR] [--ncores NCORES] [--seqlib SEQLIB]

Filter BLAT recap table based on SDFE and maxHomology.

positional arguments:
  in_dir          Input directory.
  out_dir         Output directory.

flags:
  -h, --help      show this help message and exit

optional arguments:
  -f, --fin FIN   Input file name.
                  [default: merged_table.sdfe.tsv]
  -o, --out OUT   Output file name.
                  [default: merged_table.filtered.sdfe.tsv]
  --ht HT         Homology threshold.
                  [default: 0.75]
  -s, --st ST     Self-dimerization free energy threshold (positive).
                  [default: -5]
  --sl SL         Path to seq_lib.R
                  [default: ./seq_lib.R]
</code></pre>
<h3>
<a id="user-content-8-calculate-hdfe" class="anchor" href="#8-calculate-hdfe" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>8. Calculate hdfe</h3>
<p>To calculate the hetero-dimerization free energy (hdfe) of every pair of filtered barcode candidate, use the <code>dfeCalc</code> script (based on code by Dr. Erik Wernersson).</p>
<pre><code>usage: dfeCalc [-h] -i input -o output -m mode [-l seqLen][-v]

Mandatory arguments:
-i  Input file
-o  Output file
-m  Mode. 1: reverse, 2: complement, 3: 1+2

Optional arguments:
-l  Sequence length, default: 20
-v  Verbose mode.
-h  Show this help page

</code></pre>
<p>Using the <code>-v</code> option to trigger verbose mode and have more details on the free energy calculations. The three different modes are:</p>
<ol>
<li>Calculate the HDFE of every oligo and the reverse of every other.</li>
<li>Calculate the HDFE of every oligo and the reverse complement of every other.</li>
<li>Keep the minimum (strongest bind) HDFE from the other two modes.</li>
</ol>
<p>The script <strong>input file</strong> is a oneline fasta file with the sequences from the filtered table generated at step 7. This file can be produced using the following command line:</p>
<pre><code>cat merged_table.filtered.sdfe.tsv | cut -f 4 | sed 1d | tr -d "\n" &gt; sequence.filtered.oneline.fa
</code></pre>
<h3>
<a id="user-content-9-filter-hdfe" class="anchor" href="#9-filter-hdfe" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>9. Filter (hdfe)</h3>
<p>Filter the HDFE table using the <code>hdfe_filter.R</code> script. The default threshold is set at -9 kcal/mol.</p>
<pre><code>usage: hdfe_filter.R [--help] in_dir out_dir
       [--fin FIN][--out OUT][--pmc PMC][--ht HT]

Filter BLAT recap table based on SDFE and maxHomology.

positional arguments:
  in_dir            Input directory.
  out_dir           Output directory.

flags:
  -h, --help        show this help message and exit

optional arguments:
  -f, --fin FIN     Input file name.
                    [default: hdfe.dat]
  -o, --out OUT     Output file name.
                    [default: hdfe_filtered.dat]
  -p, --pmc PMC     Output PMC file name.
                    [default: hdfe_filtered_graph.dat]
  --ht HT           Homology threshold (positive).
                    [default: -9]
</code></pre>
<h3>
<a id="user-content-10-select-barcode" class="anchor" href="#10-select-barcode" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>10. Select barcode</h3>
<p>Finally, select a set of compatible barcodes with the <a href="https://www.cs.purdue.edu/homes/dgleich/codes/maxcliques/README.html"><code>pmc</code> algorithm</a>.</p>
<pre><code>./pmc -f hdfe_filtered_graph.dat -t ncores &gt; pmc.log
</code></pre>
<pre><code>Usage: ./pmc -a alg -f graphfile -t threads -o ordering -h heu_strat -u upper_bound -l lower_bound -r reduce_wait_time -w time_limit 
  -a algorithm                 : Algorithm for solving MAX-CLIQUE: 0 = full, 1 = no neighborhood cores, 2 = only basic k-core pruning steps  
  -f graph file                : Input GRAPH file for computing the Maximum Clique (matrix market format or simple edge list). 
  -o vertex search ordering    : Order in which vertices are searched (default = deg, [kcore, dual_deg, dual_kcore, kcore_deg, rand]) 
  -d decreasing order          : Search vertices in DECREASING order. Note if '-d' is not set, then vertices are searched in increasing order by default. 
  -e neigh/edge ordering       : Ordering of neighbors/edges (default = deg, [kcore, dual_deg, dual_kcore, kcore_deg, rand]) 
  -h heuristic strategy        : Strategy for HEURISTIC method (default = kcore, [deg, dual_deg, dual_kcore, rand, 0 = skip heuristic]) 
  -u upper_bound               : UPPER-BOUND on clique size (default = K-cores).
  -l lower_bound               : LOWER-BOUND on clique size (default = Estimate using the Fast Heuristic). 
  -t threads                   : Number of THREADS for the algorithm to use (default = 1). 
  -r reduce_wait               : Number of SECONDS to wait before inducing the graph based on the unpruned vertices (default = 4 seconds). 
  -w time_limit                : Execution TIME LIMIT spent searching for max clique (default = 7 days) 
  -k clique size               : Solve K-CLIQUE problem: find clique of size k if it exists. Parameterized to be fast. 
  -s stats                     : Compute BOUNDS and other fast graph stats 
  -v verbose                   : Output additional details to the screen. 
  -? options                   : Print out this help menu.
</code></pre>
</article></body></html>